substrate/runtime/contract/src/vm.rs

```diff
@@ -0,0 +1,785 @@
+	ext: &'a mut E,
+	config: &'a Config<T>,
+	memory: sandbox::Memory,
+	gas_meter: &'a mut GasMeter<T>,
+	special_trap: Option<SpecialTrap>,
+}
+impl<'a, T: Trait, E: Ext<T> + 'a> Runtime<'a, T, E> {
+	fn memory(&self) -> &sandbox::Memory {
+		&self.memory
+	}
+	/// Save a data buffer as a result of the execution.
+	///
+	/// This function also charges gas for the returning.
+	///
+	/// Returns `Err` if there is not enough gas.
+	fn store_return_data(&mut self, data: Vec<u8>) -> Result<(), ()> {
+		let data_len = <T::Gas as As<u64>>::sa(data.len() as u64);
+		let price = (self.config.return_data_per_byte_cost)
+			.checked_mul(&data_len)
+			.ok_or_else(|| ())?;
+
+		match self.gas_meter.charge(price) {
+			GasMeterResult::Proceed => {
+				self.special_trap = Some(SpecialTrap::Return(data));
+				Ok(())
+			}
+			GasMeterResult::OutOfGas => Err(()),
+		}
+	}
+}
+
+fn to_execution_result<T: Trait, E: Ext<T>>(
+	runtime: Runtime<T, E>,
+	run_err: Option<sandbox::Error>,
+) -> Result<ExecutionResult, Error> {
+	let mut return_data = Vec::new();
+
+	// Check the exact type of the error. It could be plain trap or
+	// special runtime trap the we must recognize.
+	match (run_err, runtime.special_trap) {
+		// No traps were generated. Proceed normally.
+		(None, None) => {}
+		// Special case. The trap was the result of the execution `return` host function.
+		(Some(sandbox::Error::Execution), Some(SpecialTrap::Return(rd))) => return_data = rd,
+		// Any other kind of a trap should result in a failure.
+		(Some(_), _) => {
+			return Err(Error::Invoke);
+		}
+		_ => {
+			// All possible cases should have been handled.
+			// If the control flow reached here, then it is a logic error.
+			unreachable!();
+		}
+	}
+
+	Ok(ExecutionResult {
+		return_data,
+	})
+}
+
+/// The result of execution of a smart-contract.
+#[derive(PartialEq, Eq)]
+#[cfg_attr(feature = "std", derive(Debug))]
+pub struct ExecutionResult {
+	pub return_data: Vec<u8>,
+}
+
+impl ExecutionResult {
+	/// The result produced by the execution of the contract.
+	///
+	/// The contract can designate some buffer at the execution time via a special function.
+	/// If contract called this function with non-empty buffer it will be copied here.
+	///
+	/// Note that gas is already charged for returning the data.
+	pub fn return_data(&self) -> &[u8] {
+		&self.return_data
+	}
+}
+
+/// Execute the given code as a contract.
+pub fn execute<'a, T: Trait, E: Ext<T>>(
+	code: &[u8],
+	ext: &'a mut E,
+	gas_meter: &mut GasMeter<T>,
+) -> Result<ExecutionResult, Error> {
+	// TODO: Receive data as an argument
+
+	// ext_gas(amount: u32)
+	//
+	// Account for used gas. Traps if gas used is greater than gas limit.
+	//
+	// - amount: How much gas is used.
+	fn ext_gas<T: Trait, E: Ext<T>>(
+		e: &mut Runtime<T, E>,
+		args: &[sandbox::TypedValue],
+	) -> Result<sandbox::ReturnValue, sandbox::HostError> {
+		let amount = args[0].as_i32().unwrap() as u32;
```

Comment:
We try to avoid `unwrap` whenever possible? Can't we trap this with `?` or react with an explicit Error _should_ that happen?

Reply:
> We try to avoid `unwrap` whenever possible?

That's true! 

This is actually a longstanding issue, that present in [the current version](https://github.com/paritytech/polkadot/blob/2ecd934316e0d0c4cc34ca66c533c243768fe892/substrate/runtime/contract/src/lib.rs#L215).

The reason for that that I didn't come up with a nice way of checking signatures, and this is still blocked on JS side of things, that we still need to solve.

Since you're not first who asking [for this](https://github.com/paritytech/polkadot/pull/253#discussion_r198516511) I think it is a high time to add written comment about this : )

Reply:
at the very least we should use `expect("first arg is i32; guaranteed by macro; qed")`

Reply:
@rphmeier 

Yeah, that is the thing: for now there is no check (this is the thing we need to research on JS side) and it isn't guaranteed by anything. So `expect("shit can really happen; qed")` will look quite hilarious : ) 

On the other hand, apparently `.unwrap`s are quite visible and there is no chance of accidently missing them on reviews.

However, I'm not sure about this myself, and can change it to `.expect` if you want!